bin/fetch-cards-pdf/main.go

```diff
@@ -6,135 +6,221 @@ import (
+	flag.IntVar(&cfg.workers, "workers", 10, "number of cards to process in parallel")
+	flag.BoolVar(&cfg.secondaryOnly, "secondary-only", false, "disable download of the first cards")
 	flag.Parse()
 
 	log.Info("starting")
+	imagick.Initialize()
+	defer imagick.Terminate()
 
-	log.Debug("retrieving database")
-	res, err := http.Get(baseURL)
+	log.Debug("connecting to database")
+	db, err := sqlx.Connect("mysql", cfg.dsn)
 	if err != nil {
-		log.Error("retrieving database", "err", err)
+		log.Error("connecting to database", "err", err)
 		os.Exit(1)
 	}
 
-	log.Debug("parsing database")
-	doc, err := goquery.NewDocumentFromReader(res.Body)
-	if err != nil {
-		log.Error("reading database", "err", err)
-		os.Exit(1)
-	}
+	// create job queue
+	jobQueue := make(chan DownloadJob)
 
-	log.Debug("looking for refs")
-	type Ref struct {
-		ID      string
-		Faction string
-		Job     string
-		Name    string
-	}
-	references := make(chan Ref)
-	go func() {
-		defer close(references)
-		doc.Find("carditem").EachWithBreak(func(_ int, s *goquery.Selection) bool {
-			references <- Ref{
-				ID:      s.AttrOr(":card", ""),
-				Faction: s.AttrOr("faction", ""),
-				Job:     s.AttrOr("job", ""),
-				Name:    s.AttrOr("title", ""),
-			}
-			return true
-		})
-	}()
+	// queue front cards for all refs
+	if !cfg.secondaryOnly {
+		go queueFirstCards(cfg.ppURL, jobQueue)
+	}
 
-	log.Debug("initializing imagick")
+	// queue front cards for special cases (colossal, character unit, dragoon)
+	go queueSpecialCaseCards(db, jobQueue)
 
-	imagick.Initialize()
-	defer imagick.Terminate()
+	// queue front cards for attachments (makeda & exalted court)
+	go queueAttachmentCards(db, jobQueue)
 
+	// process jobQueue
 	var wg sync.WaitGroup
-	for w := 0; w < workers; w++ {
+	for w := 0; w < cfg.workers; w++ {
 		wg.Add(1)
-		go func() {
+		go func(w int) {
 			defer wg.Done()
 			log.Debug("starting worker", "w", w)
 
-			mw := imagick.NewMagickWand()
-			defer mw.Destroy()
-
-			for ref := range references {
-				log := log.New("ref", ref.ID, "name", ref.Name)
-				log.Debug("retrieving card")
-
-				res, err := http.Get(fmt.Sprintf("%s/?card_items_to_pdf=$%s,1", baseURL, ref.ID))
-				if err != nil {
-					log.Error("retrieving card", "err", err)
+			for job := range jobQueue {
+				if err := downloadCard(cfg.ppURL, cfg.destDir, job.RefID, job.Index); err != nil {
+					log.Error("extract card", "err", err)
 					continue
 				}
+			}
+		}(w)
+	}
+	wg.Wait()
+}
 
-				raw, err := ioutil.ReadAll(res.Body)
-				if err != nil {
-					log.Error("reading card", "err", err)
-					continue
-				}
+func queueFirstCards(ppURL string, jobQueue chan DownloadJob) {
+	log.Debug("retrieving PP cards database")
+	res, err := http.Get(ppURL)
+	if err != nil {
+		log.Error("retrieving pp cards database", "err", err)
+		os.Exit(1)
+	}
 
-				mw.Clear()
+	log.Debug("parsing PP cards database")
+	doc, err := goquery.NewDocumentFromReader(res.Body)
```

Comment:
We could probably stop doing this and only use the database, so we would avoid downloading extra cards for not-yet handled references? Also, it would probably be way faster.

Reply:
with this `Indices []int` would makes more sense, but we will do that later